---
title: "EDA of datasets"
author: "Samuel Leung"
date: "2025-03-28"
output: html_document
---

# Exploratory Data Analysis

Short RMarkdown file to briefly preprocess data

```{r load in data, message=FALSE, warning=FALSE}
library(here)
library(tidyverse)
library(ggplot2)
library(Seurat)

# read in training and test data
train <- readRDS(here("data", "UROMOL_TaLG.teachingcohort.rds"))
test <- readRDS(here("data", "knowles_matched_TaLG_final.rds"))

# isolate expression data from metadata
train.expr <- train$exprs
test.expr <- test$exprs
train.metadata <- train %>% select(-exprs)
test.metadata <- test %>% select(-exprs)
```

```{r metadata column sync}
# columns that confound prediction of recurrence
confound <- c("Progression", "PFS_time.", "Recurrence", "RFS_time", "FUtime_days.")

# columns that are irrelevant
irr <- c("UROMOL.ID", "Tumor.stage", "Tumor.grade")

# keep only columns between both train and test sets
keep <- intersect(colnames(test.metadata), colnames(train.metadata %>% select(-c(confound, irr))))
print("Common metadata columns to keep: ")
keep 
```

```{r metadata missing value analysis}
# check for missing data in the response variable `Recurrence`
# do this for both test and train sets
train.missing_response <- train.metadata$Recurrence %>% is.na()
train.metadata <- train.metadata[!train.missing_response,]
train.expr <- train.expr[!train.missing_response,]

test.missing_response <- test.metadata$Recurrence %>% is.na()
test.metadata <- test.metadata[!test.missing_response,]
test.expr <- test.expr[!test.missing_response,]

# out of the subsetted variables: check which ones have NAs
print("Preprocessing Training Data summary:")
train.metadata[,keep] %>% is.na() %>% colMeans() %>% round(2)

# Check for test set as well
print("Preprocessing Testing Data Summary")
test.metadata[,keep] %>% is.na() %>% colMeans() %>% round(2)

# only 1% of cells in Age are missing from training set. Imputing NA's by mean age for simplicity
train.metadata <- train.metadata %>%
  mutate(Age = if_else(is.na(Age), round(mean(Age, na.rm=TRUE), 0), Age))
```
```{r analyze-rfs}
# look at missing rfs data in train set
train.metadata$RFS_time %>% is.na() %>% sum()

# look at missing rfs data in test set
test.metadata$RFS_time %>% is.na() %>% sum()

# impute missing RFS using FU time
test.metadata$RFS_time[is.na(test.metadata$RFS_time)] = test.metadata %>% 
  filter(is.na(RFS_time)) %>%
  mutate(RFS_time = round(FUtime_days./30)) %>%
  select(RFS_time) %>%
  pull()

is.na(test.metadata$RFS_time)
```


```{r expr data analysis}
# find common genes between train and test set
keep.genes <- intersect(colnames(train.expr), colnames(test.expr))

# check for missing values in gene sets
print("Total number of missing values in training set: ")
train.expr[,keep.genes] %>% is.na() %>% sum()

print("Total number of missing values in testing set: ")
test.expr[,keep.genes] %>% is.na() %>% sum()
```

No missing gene expressions (Yippee!)

However, the training set is from RNA-seq data and the test set is microarray data. From a quick glance of the data, it seems like the data are from separate distributions

```{r train-test glance}
print("Train set quick view")
train.expr[1:5, keep.genes[1:5]]

print("Test set quick view")
test.expr[1:5, keep.genes[1:5]]
```

Given this issue, normalization with metrics from the train set will most likely not be comparable to those of the test set. We opt to log-transform the expression data to make the two datasets more comparable:

```{r log train-test glance}
print("Train set quick view after log transform")
log(train.expr[1:5, keep.genes[1:5]] + 1)

print("Test set quick view after log transform")
log(test.expr[1:5, keep.genes[1:5]] + 1)
```

## Clustering

We want to see how well gene expression can separate the data, so we perform unsupervised clustering on the training set:

```{r train-set-cluster-and-plot, warning=FALSE}
# create seurat object
seurat <- CreateSeuratObject(counts = t(train.expr[,keep.genes]))

# log-normalize
seurat <- NormalizeData(seurat, verbose = FALSE)

# find variable features + scale data
seurat <- FindVariableFeatures(seurat, verbose = FALSE)
seurat <- ScaleData(seurat, verbose = FALSE)

# Perform PCA + UMAP + Leiden Clustering
seurat <- RunPCA(seurat, verbose = FALSE)
seurat <- RunUMAP(seurat, dims = 1:20, verbose = FALSE)
seurat <- RunTSNE(seurat, dims = 1:20, verbose = FALSE)
seurat <- FindNeighbors(seurat, verbose = FALSE)
seurat <- FindClusters(seurat, algorithm = 4, random.seed = 124, verbose = FALSE)
```

```{r View Clusters}
# view raw clustering
DimPlot(object = seurat, reduction = "umap")
DimPlot(object = seurat, reduction = "tsne")

# view clustering with recurrence labels
seurat$Recurrence = ifelse(train.metadata$Recurrence, "Recurrence", "No Recurrence")
DimPlot(object = seurat, reduction = "umap", group.by = "Recurrence")
DimPlot(object = seurat, reduction = "tsne", group.by = "Recurrence")

# view clustering based on EAU risk
seurat$EAU = train.metadata$EAU.risk
DimPlot(object = seurat, reduction = "umap", group.by = "EAU")
DimPlot(object = seurat, reduction = "tsne", group.by = "EAU")

# view clustering based on UROMOL ID
seurat$UROMOL = train.metadata$UROMOL2021.classification
DimPlot(object = seurat, reduction = "umap", group.by = "UROMOL")
DimPlot(object = seurat, reduction = "tsne", group.by = "UROMOL")
```

## Observations from unsupervised clustering

-   Leiden clusters do not re-capitulate recurrence, EAU risk, or UROMOL identification clusters

### Save filtered train + test sets

```{r save-filtered-sets}
# combine metadata and expr datasets by UROMOL id
# also add back response variable
train.to_save <- merge(train.metadata[,c("Recurrence","RFS_time",keep)], train.expr[,keep.genes], by = "row.names")
test.to_save <- merge(test.metadata[,c("Recurrence","RFS_time",keep)], test.expr[,keep.genes], by = "row.names")

# save to data directory
saveRDS(train.to_save, file = here("data", "filtered_train_set.rds"))
saveRDS(test.to_save, file = here("data", "filtered_test_set.rds"))
```

